[ThatGirlJam] iP

ThatGirlJam opened the pull request:
# This is a pull request to Duke with my version -- Duchess.

Some things you can do with it 
- Make a List
- Mark items 
- Schedule deadlines and events

For example
1. [T] [ ] Make my bed
2. [D] [ ] Finish iP (by: Week 6)

Duchess runs with *this* method : 
```java
public static void main(String[] args) {
        try {
            new Duchess().run();
        } catch (DuchessException e) {
            System.out.println(e.getMessage());
        }

    }
```

**Imagine** a task list like 
- [x] Make my bed
- [ ] Finish iP (by: Week 6)

:+1:

> I heard this is a ~~good~~ great chatbot! - Jaime 

Access Duchess here : 
[https://github.com/nus-cs2103-AY2324S2/ip/pull/305](url)

`System.out.println("Get Duchess now!")`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ThatGirlJam committed:
Rename Duke to Duchess

ThatGirlJam committed:
Add greeting and exit skeleton

ThatGirlJam committed:
refactor: separate greetings into respective functions

ThatGirlJam committed:
Print an echo of any input given, exit if input is bye.

ThatGirlJam committed:
Remove echo function. Add every input to a task list and display list when prompted.

ThatGirlJam committed:
Add Task class with getStatusIcon

ThatGirlJam committed:
refactor: change tasks in Duchess to be represented as Task

ThatGirlJam committed:
fix: make task description and not id be listed when printList

ThatGirlJam committed:
Add unmarkAsDone and markAsDone to Task class

ThatGirlJam committed:
Print box for status

ThatGirlJam committed:
refactor: change echo from if to switch

ThatGirlJam committed:
Add mark and unmark function to Duchess class

ThatGirlJam committed:
Add Deadline class and toString for Deadline and Task class

ThatGirlJam committed:
Add ToDo task with toString method

ThatGirlJam committed:
Add Event class with toString method

ThatGirlJam committed:
refactor: made Duchess sound more polite

ThatGirlJam committed:
Add tracking for todo, event, deadline. Made Event more generic start and end.

ThatGirlJam committed:
refactor: add comments for new adding of ToDo, Deadline, Event

ThatGirlJam committed:
fix: remove unnecessary whitespaces and added a test case

ThatGirlJam committed:
Add test inputs and outputs for level 0 to 4

ThatGirlJam committed:
refactor: change error messages to fit Duchess personality

ThatGirlJam committed:
Handle exceptions by printing error messages

ThatGirlJam committed:
Add a delete function for tasks, changed to ArrayList

ThatGirlJam committed:
Add enum for task type, prepare for future use

ThatGirlJam committed:
Change methods that work on instances to non-static and update input output tests

There were methods that were set as static even though they shouldn't be as they deal with object instances. Change those.

Update the tests to have the specific types of tasks.

ThatGirlJam committed:
Add constructors to Deadline, Event, Task and ToDo for tasks already with an isDone state

ThatGirlJam committed:
Save the tasks in the hard disk automatically whenever the task list changes and load the data when the chatbot starts up

The file name and relative parth is hard coded as ./data/duchess.txt

The tasks are saved such that each task in the list is iterated through to be updated.

ThatGirlJam committed:
Merge branch 'branch-Level-7'
Save data every time list changes and load data when chatbot starts up.

ThatGirlJam committed:
Change deadline to type LocalDateTime so that it can understand dates and times .

Deadline accepts DD-MM-YYYY HHmm and prints as MMM dd yyyy hh:mm a.

ThatGirlJam committed:
Merge branch 'branch-Level-8'
Store deadline dates as a java time LocalDateTime and print in a different format.

ThatGirlJam committed:
Extract task type enum to separate class

ThatGirlJam committed:
Add TaskList and pause storing/loading of data.

ThatGirlJam committed:
Add Storage class and incorporate both TaskList and Storage into Duchess

ThatGirlJam committed:
Add Parser class that parses data to load

ThatGirlJam committed:
Add Parser class that parses data to load

ThatGirlJam committed:
Add Ui class to handle interface interactions

ThatGirlJam committed:
Move input handling to Ui

ThatGirlJam committed:
Refactor code to make more OOP

ThatGirlJam committed:
Merge branch 'A-moreOOP'
Extract the Storage, Parser, Ui and TaskList classes from Duchess.

ThatGirlJam committed:
Repackage classes into duchess and duchess.task packages

ThatGirlJam committed:
Merge branch 'branch-a-packages'
Repackage the classes into duchess and duchess.task packages.

ThatGirlJam committed:
Merge branch 'add-gradle-support'
Add gradle support from Duke repo.

ThatGirlJam committed:
Add TaskListTest and DeadlineTest

ThatGirlJam committed:
Add addToDoTest, deleteTaskTest, deadlineToStringTest

ThatGirlJam committed:
Merge branch 'branch-a-junit'
Add the JUnit tests for deadlineToString, deleteTaskTest, addToDoTest

ThatGirlJam committed:
Add .gradle to be ignored

ThatGirlJam committed:
Merge branch 'branch-a-jar'
Add .gradle to the gitignore

ThatGirlJam committed:
Add Java Doc to the files

ThatGirlJam committed:
Add line to java doc for Event

ThatGirlJam committed:
Refactor code to fit coding standard, such as when using switch case

ThatGirlJam committed:
Refactor code such that JavaDoc fits the coding standard

ThatGirlJam committed:
Add function of finding the list of items that match a keyword in its description, showing their indexes from the original list

ThatGirlJam committed:
Refactor code to adhere to coding standard. Solve conflict in merge.

ThatGirlJam committed:
Add find by keyword function, and include finTasksByKeyWord javaDoc

KinTatHo commented on src/main/java/duchess/Parser.java:
```diff
@@ -0,0 +1,48 @@
+        boolean isDone = parts[1].trim().equals("1");
+        String description = parts[2].trim();
+
+        switch (type) {
```
Good job remembering about the switch not having indentation

KinTatHo commented on src/main/java/duchess/TaskList.java:
```diff
@@ -0,0 +1,241 @@
+ * It provides methods to add, delete, and manipulate tasks in the list.
+ */
+public class TaskList {
+    private static final int MAX_TASKS = 100;
```
Nice job remembering!

KinTatHo commented on src/main/java/duchess/Parser.java:
Maybe you can store this parser into a parser package?

ThatGirlJam replied:
Thank you! Will do 👍

KinTatHo commented on src/main/java/duchess/Ui.java:
Similar to parser, how about a ui package?

ThatGirlJam replied:
Thank you! Will do that 👍

KinTatHo reviewed (commented):
Overall, good quality and understandable code. Unable to spot any severe mistakes so far

ThatGirlJam committed:
Packaged Parser, Storage and Ui into their own packages

ThatGirlJam committed:
Add config files for checkstyle and make necessary style changes

ThatGirlJam committed:
Merge branch 'branch-A-CheckStyle'
Add the config files for checkstyle and do the necessary style changes to the code"

iamtr commented on src/main/java/duchess/TaskList.java:
```diff
@@ -0,0 +1,247 @@
+     * Prints the task list.
+     */
+    public void printTaskList() {
+        //printHorizontalLine();
```
Should this unused call be removed, or will it be use in the future?

iamtr reviewed (commented):
Overall, I felt like your PR is amazing! Code was well structured and very readable.

 There were not much things to comment, just some nits on deleting unused code.

ThatGirlJam committed:
Add launcher that allows entering text and scrolling of iteration 1

ThatGirlJam committed:
Add images and heard from the chatbot for iteration 2

ThatGirlJam committed:
Refactor code to make chat bot on left and user on right for iteration 3

ThatGirlJam committed:
Refactor code to use fxml in handling GUI launch

ThatGirlJam committed:
Add command recognition in GUI and relevant GUI responses

ThatGirlJam committed:
Change default images to the duchess and knight images

ThatGirlJam committed:
Refactor tests to fit new GUI duchess

ThatGirlJam committed:
Add javadocs and adhere to checkstyle

ThatGirlJam committed:
Add alert to say bye before closing the main window to ensure saving of task list

ThatGirlJam committed:
Refactor code to fit checkstyle

ThatGirlJam committed:
Merge branch 'branch-Level-10'
Add Main, MainWindow, DialogBox and Launcher to allow a GUI format for Duchess

Duchess no longer supports text input ui, and only does GUI format. The main window is opened when start launcher and can only be closed when user says bye to Duchess.

Added images to represent user and Duchess.

ThatGirlJam committed:
Change shadowjar file name

ThatGirlJam committed:
Add assertions to ensure proper exit and saving before MainWindow closes

Added assertions about making sure Duchess exits only if it isRunning, as it cannot exit again if it is already not running.

Made sure Duchess restarts if a new command is written even after writing bye to Duchess.

ThatGirlJam committed:
Refactor Duchess and MainWindow to be more readable

ThatGirlJam committed:
Add assert to ensure user and duchess images are not null before initialising dialog box

ThatGirlJam committed:
Refactor MainWindow, TaskList, Parser and Storage to be more readable

ThatGirlJam committed:
Add opening greeting to Duchess on launch

ThatGirlJam committed:
Add comment on stream use in MainWindow

ThatGirlJam committed:
Merge pull request #1 from ThatGirlJam/branch-A-CodeQuality

Branch a code quality

ThatGirlJam committed:
Resolve merge conflicts in Duchess, MainWindow and TaskList

ThatGirlJam committed:
Merge pull request #3 from ThatGirlJam/branch-A-Assertions

Branch a assertions

ThatGirlJam committed:
Merge pull request #2 from ThatGirlJam/branch-A-Streams

Add comment on stream use in MainWindow

ThatGirlJam committed:
Add CI with a gradle.yml file

ThatGirlJam committed:
Merge branch 'branch-A-CI'
Add gradle.yml file into .github/workflows folder

ThatGirlJam committed:
Refactor code to use guard conditions

ThatGirlJam committed:
Add listDuplicateTasks to allow command duplicates to return a list of all the duplicated tasks that were added the latest

ThatGirlJam committed:
Merge branch 'branch-BCD-Extension'
Add the function of the command 'duplicates' allowing a list of the last added duplicated tasks to be returned

Refactored some code as well to show subList.

ThatGirlJam committed:
Refactor code to pass checkstyle

ThatGirlJam committed:
Refactor header comments to match Java Standard

ThatGirlJam committed:
Add HBox to contain the text field and send button for better alignment

ThatGirlJam committed:
Add different colours for duchess and user text and allow for no minimum height for texts

ThatGirlJam committed:
Crop images to be rounded and round the text

ThatGirlJam committed:
Refactor code to make greetings as constants

ThatGirlJam committed:
Change greetings to be more align with the duchess theme

ThatGirlJam committed:
Catch error in case data file is in wrong format

ThatGirlJam committed:
Add markTaskAsDoneTest and unmarkTaskAsDoneTest to test more task list methods

ThatGirlJam committed:
Add findTasksByKeywordTest to check if finding tasks by keyword is accurate

ThatGirlJam committed:
Add a title to the main window of the gui

ThatGirlJam committed:
Change size of pictures for DaDuchess and DaUser

ThatGirlJam committed:
Add picture of the Ui to the README document

ThatGirlJam committed:
Fix screenshot of Ui not showing up on docs

ThatGirlJam committed:
Add the features and how to use them to the README

ThatGirlJam committed:
Change width of screenshot of the Ui picture for README User Guide

ThatGirlJam committed:
Add bolding to format descriptions in each feature command

ThatGirlJam committed:
Change picture to be able to be seen

ThatGirlJam committed:
Refactor User Guide to change style of format descriptions

ThatGirlJam committed:
Add list of all features into user guide

ThatGirlJam committed:
Change styling for format for commands

ThatGirlJam committed:
Change style for command formats to change brackets so full format can be seen

ThatGirlJam committed:
Refactor code to fit checkstyle standard

ThatGirlJam committed:
Change shadow jar file name to duchess release

ThatGirlJam committed:
Fix hardcoded data paths and fix jar generation

ThatGirlJam committed:
Update name of shadowjar archive base name to duchessrelease2

ThatGirlJam committed:
Update name of shadowjar archive base name to duchessversion2
